### PR TITLE
ALSA: hda: Unify codec construction

### PR DESCRIPTION
**Pull request by crojewsk-intel**
A follow up to the recent HDAudio fixes series [1]. Given the recently reported regression [2], before the page fault occurring on codec shutdown can be fixed, codec construction procedure needs to be updated for skylake and sof-intel drivers. Drivers: pci-hda and avs need no changes - already making use of snd_hda_codec_device_init().

Changes in v2:
- drop snd_hda_ext_core <-> snd_hda_codec dependency by calling snd_hda_codec_device_init() directly in skylake and sof drivers probe enumeration routines, as suggested by Takashi
- skylake/sof portion of the change have been split into two separate patches
- additional patch cleaning up unused hdac_ext code provided
- note: first patch in the series unchanged so Mark's ack remained

1: https://lore.kernel.org/alsa-devel/20220706120230.427296-7-cezary.rojewski@intel.com/
2: https://lore.kernel.org/alsa-devel/3c40df55-3aee-1e08-493b-7b30cd84dc00@linux.intel.com/

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by crojewsk-intel** [comment]
For reference, v1 of the series can be found [here](https://lore.kernel.org/alsa-devel/20220720130622.146973-1-cezary.rojewski@intel.com/).

**Comment by plbossart** [comment]
Which of these two GitHub handles should be added to the SOF project? 
@crojewsk-intel and @crojewsk

**Comment by crojewsk-intel** [comment]
> Which of these two GitHub handles should be added to the SOF project? @crojewsk-intel and @crojewsk

If I'm not mistaken, @crojewsk already is. This account is here as Intel requests RW rights and I don't want to lose my Warcraft III stuff lol.

**Comment by crojewsk-intel** [comment]
Perhaps both if that's not the problem?

File: sound/soc/intel/skylake/skl.c
Comment (crojewsk-intel) [inline sound/soc/intel/skylake/skl.c]: Converting to `static` in next update.

File: sound/soc/sof/intel/hda-codec.c
Comment (crojewsk-intel) [inline sound/soc/sof/intel/hda-codec.c]: Converting to `static` in next update.

File: sound/soc/sof/intel/hda-codec.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-codec.c]: how does this work?

hda_priv is initialized with zeroes, so hda_priv->codec is NULL.

Comment (crojewsk-intel) [reply sound/soc/sof/intel/hda-codec.c]: This line is later removed. The very first patch changes field to pointer, literally. It does not care about consequences - these are addressed by the following patches.

Comment (plbossart) [reply sound/soc/sof/intel/hda-codec.c]: wouldn't this break git bisect?

Comment (crojewsk-intel) [reply sound/soc/sof/intel/hda-codec.c]: Compile-time no. Runtime yes. If the patch is to make things work, we would end up with 1,2,3 squashed up together i.e. one big patch. I decided to go with smaller patches and add: "Preliminary step (...)" in the commit message instead.

Comment (plbossart) [reply sound/soc/sof/intel/hda-codec.c]: if we can't use git bisect to run tests, that's a deal breaker. Sorry.

Comment (crojewsk-intel) [reply sound/soc/sof/intel/hda-codec.c]: Will see what I can do. Perhaps I'll define the new functions first, then do field -> pointer + skl & sof update in single following patch.

Comment (plbossart) [reply sound/soc/sof/intel/hda-codec.c]: Thanks. git bisect is heavily used to root-cause upstream regressions or side effects of changes in programming sequences, I always insist that we keep the functionality even when doing open-heart surgery in the code.

**Comment by crojewsk-intel** [comment]
Update. Raw delta remained 1:1 to previous version +113, -134. A lot of code shuffling between the first three patches though.

The following has been done to make git bisect happy:
- new functions that aim to replace hdac_ext codec init & exit functionality are added first - for skylake and sof drivers both
- third patch now combines the "field -> pointer" change for `hdac_hda_priv->codec` plus the codec-enumeration adjustments for skylake and sof drivers

Dropped Mark's ack as the code changed too much. Last three patches remained basically unchanged - `ALSA: hda: Remove codec init and exit routines` now removes `hdac_to_hda_priv()` macro on top its previous cleanups.

**Comment by plbossart** [comment]
CI tests don't seem too good @crojewsk-intel ? see https://sof-ci.01.org/linuxpr/PR3775/build445/devicetest/

File: sound/soc/sof/intel/hda-codec.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda-codec.c]: Hmm, not sure if the type setting goes correctly here. It seems the assignment on L175 is no longer used and variable "type" goes unused after this change.

In SOF, we basicly only use HDA_DEV_ASOC to support one old configuration (some GLK chromebooks using hdac_hdmi), but in 97% of the configurations, init should be done with HDA_DEV_LEGACY (and use hdac_hda.c).

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @crojewsk for digging into this! The cleanup seems good, but I have one concern about the SOF patch to move to new init. Please see comments inline.

File: sound/soc/sof/intel/hda-codec.c
Comment (crojewsk-intel) [reply sound/soc/sof/intel/hda-codec.c]: Good catch. This is probably the cause for the CI failing. Updated `hda_codec_device_init()` to accept additional 'type' argument to address the problem. For skylake-driver it's easier - it was always assigning `HDA_DEV_ASOC` upfront.

**Comment by crojewsk-intel** [comment]
> CI tests don't seem too good @crojewsk-intel ? see https://sof-ci.01.org/linuxpr/PR3775/build445/devicetest/

Indeed! Kai helped me in understanding the logs. Update on its way.

**Comment by crojewsk-intel** [comment]
Results look much better now - except for 2x timeout. Guess problem with a platform? Happens a lot in my farm too..

File: sound/pci/hda/hda_codec.c
Comment (kv2019i) [inline sound/pci/hda/hda_codec.c:928]: typo in commit: s/Iinitialization/Initialization/

**Review by kv2019i — APPROVED** [review]
Thanks @crojewsk , looks good now! One minor typo in commit msg, but nothing to block merge. The few timeouts seem unrelated to your PR.

File: sound/pci/hda/hda_codec.c
Comment (crojewsk-intel) [reply sound/pci/hda/hda_codec.c:928]: Should I send an update here or can this be done when sending series to upstream?

Comment (kv2019i) [reply sound/pci/hda/hda_codec.c:928]: @crojewsk-intel Can be done when sending upstream, no problem.

**Review by plbossart — APPROVED** [review]
feel free to add my tag

Reviewed-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

Probably best to send in 2+ weeks, when the merge window closes, since it's a relatively big change impacting two maintainers.

**Comment by crojewsk-intel** [comment]
Closing as the PR's purpose has been fulfilled. I'll be sending this to upstream next Friday or so. Thanks!

**Pull request closed without merging**
